packages/opentelemetry-exporter-collector/src/transform.ts

```diff
@@ -0,0 +1,241 @@
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { hexToBase64, hrTimeToTimeStamp } from '@opentelemetry/core';
+import { ReadableSpan } from '@opentelemetry/tracing';
+import { Attributes, Link, TimedEvent, TraceState } from '@opentelemetry/types';
+import * as collectorTypes from './types';
+
+const OT_MAX_STRING_LENGTH = 128;
+const OT_MAX_ATTRIBUTES = 30;
+
+const LINK_TYPE_UNSPECIFIED: collectorTypes.LinkTypeUnspecified = 0;
+// const LINK_TYPE_CHILD_LINKED_SPAN: collectorTypes.LinkTypeChildLinkedSpan = 1;
+const LINK_TYPE_PARENT_LINKED_SPAN: collectorTypes.LinkTypeParentLinkedSpan = 2;
+
+/**
+ * convert string to maximum length of 128, providing information of truncated bytes
+ * @param name - string to be converted
+ */
+export function toCollectorTruncatableString(
+  name: string
+): collectorTypes.TruncatableString {
+  const value = name.substr(0, OT_MAX_STRING_LENGTH);
+  const truncatedByteCount =
+    name.length > OT_MAX_STRING_LENGTH ? name.length - OT_MAX_STRING_LENGTH : 0;
+
+  return { value, truncatedByteCount };
+}
+
+/**
+ * convert attributes
+ * @param attributes
+ * @param maxAttributes - default value is 30
+ */
+export function toCollectorAttributes(
+  attributes: Attributes,
+  maxAttributes: number = OT_MAX_ATTRIBUTES
+): collectorTypes.Attributes {
+  const attributeMap: collectorTypes.AttributeMap = {};
+  let droppedAttributesCount = 0;
+
+  const keys = Object.keys(attributes || {});
+
+  const countKeys = Math.min(keys.length, maxAttributes);
+  if (keys.length > maxAttributes) {
+    droppedAttributesCount = keys.length - maxAttributes;
+  }
+
+  for (let i = 0; i <= countKeys - 1; i++) {
+    const key = keys[i];
+    const eventAttributeValue = toCollectorEventValue(
+      attributes && attributes[key]
+    );
+    if (eventAttributeValue) {
+      attributeMap[key] = eventAttributeValue;
+    }
+  }
+
+  return {
+    droppedAttributesCount,
+    attributeMap,
+  };
+}
+
+/**
+ * convert event value
+ * @param value event value
+ */
+export function toCollectorEventValue(
+  value: unknown
+): collectorTypes.AttributeValue | undefined {
+  const attributeValue: collectorTypes.AttributeValue = {};
+
+  if (typeof value === 'string') {
+    attributeValue.stringValue = toCollectorTruncatableString(value);
+  } else if (typeof value === 'boolean') {
+    attributeValue.boolValue = value;
+  } else if (typeof value === 'number') {
+    if (Math.floor(value) === value) {
```

Comment:
What if the number is too larger for an int-64? What would you think about just always using `doubleValue` here since it's a direct analog to the JS `number` type.

I would think that once we support bigint, that a bigint below a certain value should become an intValue (and over that value be a string probably)

Reply:
hmm good catch and interesting, but I think I can modify this into 
`if (value < Number.MAX_SAFE_INTEGER && Math.floor(value) === value) {` this way it should convert correctly, WDYT ?

Reply:
I agree with Dave. All javascript numbers are doubles.